oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/introspection/OAuth2TokenIntrospectionClient.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.introspection;
+
+import java.util.Map;
+
+/**
+ * A client to an
+ * <a href="https://tools.ietf.org/html/rfc7662" target="_blank">OAuth 2.0 Introspection Endpoint</a>.
+ *
+ * Basically, this client is handy when a resource server authenticates opaque OAuth 2.0 tokens.
+ * It's also nice when a resource server simply can't decode tokens - whether the tokens are opaque or not -
+ * and would prefer to delegate that task to an authorization server.
+ *
+ * @author Josh Cummings
+ * @since 5.2
+ */
+public interface OAuth2TokenIntrospectionClient {
+
+	/**
+	 * Introspect the given token's attributes
+	 *
+	 * @param token the token to introspect
+	 * @return the token's attributes, including whether or not the token is active
+	 */
+	Map<String, Object> introspect(String token);
```

Comment:
Keep in mind that we will likely need to support static and dynamic custom parameters. Dynamic parameters may come from the `HttpServletRequest` or other source?

Reply:
I like the idea of returning a response domain object as it provides some greater API flexibility than returning a `Map`. 

I'm not certain that it's worth it, though, since it seems to me that it would just be a `Map`-holder. In that case, I think we'll get more mileage out of having a dedicated token + attributes class like `OAuth2TokenAttributes` when the attributes get exposed in controllers. What benefits do you see us getting from a response object?

I don't yet see the benefit of a request domain object. The reason is that `token_type_hint` only defines two values, `access_token` and `refresh_token` and it would be pretty odd, I think, for folks to be presenting a resource server with a refresh token as credentials. (I'm not predicting that folks won't want to do it - I just think that's strange enough to keep it from being a reason for us to introduce a domain object.)

If a user's authorization server takes custom `token_type_hint` values or custom request parameters, it would seem they are already in custom territory and would need to pull stuff out of the upstream `Authentication` object anyway, which means a custom authentication provider.

@rwinch, feel free to chime in as well if you have thoughts about this contract.

Reply:
The main concern I have with this API `Map<String, Object> introspect(String token)` is that it does not provide us the flexibility to build on it if we need to support custom request parameters at a later point. If we had a request object than adding another attribute will not break the core contract. Same applies to the response although a `Map` will provide more flexibility.

Have you conducted a search of well-known providers to see what their introspection endpoint supports for request parameters? I did a quick one and here is what I found.

[Okta](https://developer.okta.com/docs/api/resources/oidc/#introspect) - supports a few request parameters.

[IBM](https://www.ibm.com/support/knowledgecenter/en/SSPREK_9.0.5/com.ibm.isam.doc/config/concept/con_oauth_introspection.html) Here is what they have to say about the optional `token_type_hint`  parameter:

> The introspection endpoint supports use of the 'token_type_hint' as per section 2.1 (https://tools.ietf.org/html/rfc7662#section-2.1). This allows an optimization in the time of the token lookup. This does not limit the breadth of the search for the token in the token cache. Any token type will still be found even when its type is not the same as the hint.

[Keycloak](https://www.keycloak.org/docs/2.5/authorization_services/topics/service/protection/token-introspection.html) - They support the `token_type_hint` parameter as well but the value is `requesting_party_token` - which is used by a RP to introspect an access token

Reply:
Also, based on Okta's and Keycloak's API - the `token_type_hint` may be either `access_token`, `refresh_token`, `requesting_party_token`, maybe even `id_token` - tells me that this client might belong in `oauth2-client`. Something that we should consider at least.

Reply:
The only thing that Resource Server receives in an HTTP request is the bearer token itself, so there isn't a way for the request to indicate additional parameters to the introspection client. If custom parameters were necessary, they would likely get configured at the application level.

The majority of the parameters you indicated are related to client authentication, which can be done as a `RestTemplate` interceptor, though I think this might change to something more like what other Spring Security clients offer via `RequestEntity` `Converter`s. In that case, these would be the way that the request would be customized.

As I see it, the reason for creating a dedicated object is to provide additional request-level parameters, which Resource Server doesn't have.

That said, there are likely some legacy stories out there where clients send resource servers more than just the bearer token. Perhaps in these same stories those extra request details get transmuted to an introspection invocation. In circumstances like these, there may be an argument for a request object. Is this the kind of scenario you are referring to?

> this client might belong in `oauth2-client`

This is an interesting point because if the client goes there, then things other than Resource Server may invoke it. And in that case, I think a dedicated request object makes a lot of sense.

Reply:
I think we should wait until we have a clear use case other than Bearer Token Authentication. Or, at least, we shouldn't make the change in this PR since we are still speaking in the context of how the Resource Server would use this. There's a possibility that #6544 will shed light on other use cases. Or, I've added #6798 to make sure that we add support for custom parameters, and we may discover things there, too.

In the meantime, I think the package name is wrong. Currently, it is `org.springframework.security.oauth2.introspection`, which implies more general usage of this client, but clearly that's not the intent. We should move it to `org.springframework.security.oauth2.server.resource.introspection` until it's clear that we want to expose the client more generally.

> That said, there are likely some legacy stories out there where clients send more than just the bearer token to a resource server.

I'd like to add a bit more color to this comment I made earlier.

The `AuthenticationManager` API is abstracted away from the HTTP stack altogether. So, for an application to add more request detail to an authentication mechanism (read: a special header or request parameter -- this is like adding a special authentication header to BASIC authentication, not unheard of, but a little odd), the application needs to do two separate things: first create a custom `AuthenticationDetailsSource` and second create a custom `AuthenticationProvider` that knows how to handle their custom details.

Using that notion as a basis, we could hypothetically create an introspection request object, as you indicated, that has an `Object details` field and the `OAuth2IntrospectionAuthenticationProvider` would ferry the details from `Authentication#getDetails` to `OAuth2TokenIntrospectionRequest#setDetails` or similar. But, then the user would also need to provide a custom way to turn those details into their actual custom request.

In those cases, as a developer, I'd want to invoke the "let me talk to your manager" clause and create a custom filter that address the endpoint directly:

```java
public class MyFilter extends OncePerRequestFilter {
    protected void doFilterInternal(...) {
        String token = this.bearerTokenResolver.resolve(request);
        // ... get other things from request
        Map<String, Object> response = restTemplate.exchange(requestEntity, Map.class);
        Optional.ofNullable(response.get("active")).filter(... check if active ...)
                .map(active -> toOAuth2IntrospectionAuthenticationToken(response))
                .orElse(this.bearerTokenEntryPoint.commence(...));
    }
}
```

Or, if I didn't mind tying myself to the web tier, then I'd just use `RequestContextHolder` to get what extra details I needed.

Again, to restate, this is just what I'm thinking about for custom request-level attributes. For stuff that isn't part of the request (e.g. the resource server's IP address, the client credentials, other custom, but static, attributes), these can be added via a custom `RestOperations` or perhaps better, a custom `RequestEntity` `Converter`.